### PR TITLE
operations: Add initial support for migrating between deployment modes

### PR DESCRIPTION
**Pull request by jhesketh**
#### What this PR does

This allows an installation to deploy both microservices and the read-write backend services at the same time to put the cluster in a transitional state while migrating between the two modes.

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by pracucci — COMMENTED** [review]
I would like to see a jsonnet "test" showing the actual manifests, so we can diff it.

File: operations/mimir/config.libsonnet [hunk 1]
```diff
@@ -9,10 +9,20 @@
...
+    // `migration-read-write`: Provides both the microservices and read-write services.
+    //                         Gateway routes to read-write services.
+    //                         Expected as second state when migrating towards read-write.
     deployment_mode: 'microservices',
-    is_microservices_deployment_mode: $._config.deployment_mode == 'microservices',
-    is_read_write_deployment_mode: $._config.deployment_mode == 'read-write',
+    is_microservices_deployment_mode: $._config.deployment_mode == 'microservices' || std.startsWith($._config.deployment_mode, 'migration-'),
+    is_read_write_deployment_mode: $._config.deployment_mode == 'read-write' || std.startsWith($._config.deployment_mode, 'migration-'),
+    _gateway_service_target_microservices: $._config.deployment_mode == 'microservices' || $._config.deployment_mode == 'migration-microservices',
+    _gateway_service_target_read_write: !($._config._gateway_service_target_microservices),
```
[the remark this answers is not in the document]
Comment (jhesketh) [reply operations/mimir/config.libsonnet:25]: I need to rework how the routing side of the migration will look and likely remove these.

File: operations/mimir/config.libsonnet [hunk 2]
```diff
@@ -9,10 +9,20 @@
...
-    // The deployment mode to use. Supported values are: microservices, read-write.
+    // The deployment mode to use. Supported values are:
+    // `microservices`: Provides only the k8s objects for each component as microservices.
+    // `read-write`: Provides only mimir-read, mimir-write, and mimir-backend k8s objects.
+    // `migration-microservices`: Provides both the microservices and read-write services.
+    //                            Gateway continues routing to microservices.
+    //                            Expected as first state when migrating towards read-write.
+    // `migration-read-write`: Provides both the microservices and read-write services.
+    //                         Gateway routes to read-write services.
+    //                         Expected as second state when migrating towards read-write.
```
[the remark this answers is not in the document]
Comment (jhesketh) [reply operations/mimir/config.libsonnet:20]: I need to rework how the routing side of the migration will look and likely remove these.

File: operations/mimir/multi-zone.libsonnet [hunk 3]
```diff
@@ -26,6 +26,8 @@
     multi_zone_store_gateway_replicas: 0,
     multi_zone_store_gateway_max_unavailable: 10,
 
+    multi_zone_store_gateway_unregister_on_shutdown: false,
```
Comment (pracucci) [inline operations/mimir/multi-zone.libsonnet:29]: I would keep the same consistent with `unregister_ingesters_on_shutdown`, so `unregister_store_gateways_on_shutdown`. This feature is unrelated from "multi-zone", so no need to prefix it with `multi_zone_`.

**Review by pracucci — APPROVED** [review]
LGTM! (modulo a nit)

File: operations/mimir/multi-zone.libsonnet (hunk 3 above)
Comment (jhesketh) [reply operations/mimir/multi-zone.libsonnet:29]: After offline discussion, the challenge here is that multi-zone needs to be specifically set to something other than the default for `store-gateway.sharding-ring.unregister-on-shutdown`. This would be a regression requiring any operators using multi-zone to set `unregister_store_gateways_on_shutdown=false` manually, or otherwise still require an extra variable to control.

Comment (pracucci) [reply operations/mimir/multi-zone.libsonnet:29]: Discussed on Slack. Actually we need a different setting whether it's multi-zone or not, so let's move on with this approach.
